aten/src/ATen/core/ivalue_inl.h

```diff
@@ -346,18 +348,29 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   /**
    * Wait on the future until it completes.
    */
-  void wait() {
+  void wait(std::chrono::milliseconds timeout = kNoTimeout) {
     std::unique_lock<std::mutex> lock(mutex_);
-    finished_cv_.wait(lock, [&]() -> bool { return completed_; });
+    if (timeout == kNoTimeout) {
+      // This waits without a timeout.
+      finished_cv_.wait(lock, [&]() -> bool { return completed_.load(); });
+    } else {
+      // Waits for the user-provided timeout.
+      finished_cv_.wait_for(lock, timeout, [&]() -> bool { return completed_.load(); });
+      if (!completed_) {
+        // Throw exception if the wait operation timed out and the work was not
+        // completed.
+        throw std::runtime_error("Future::wait() timed out!");
```

Comment:
@lw has some pros/cons about throwing here, but if we do choose to do so, I'd prefer torch_check here so we can get a full cpp stack trace.